### PR TITLE
Frontend multi-language support #1690

### PR DESCRIPTION
**Pull request by bnodir**
## Purpose
This PR is to add multi-language support to the frontend of the app, which was previously only available in English. This new feature, along with i18next, will enable the app to automatically adapt to the user's browser language, thereby improving user's experience.

## Does this introduce a breaking change?

When developers merge from main and run the server, azd up, or azd deploy, will this produce an error?
If you're not sure, try it out on an old environment.

```
[ ] Yes
[x] No
```

## Does this require changes to learn.microsoft.com docs?

This repository is referenced by [this tutorial](https://learn.microsoft.com/azure/developer/python/get-started-app-chat-template)
which includes deployment, settings and usage instructions. If text or screenshot need to change in the tutorial,
check the box below and notify the tutorial author. A Microsoft employee can do this for you if you're an external contributor.

```
[x] Yes
[ ] No
```

## Type of change

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## Code quality checklist
- [ ] The current tests all pass
- [ ] I added tests that prove my fix is effective or that my feature works
- [ ] I ran `python -m pytest --cov` to verify 100% coverage of added lines
- [ ] I ran `python -m mypy` to check for type errors
- [x] I either used the pre-commit hooks or ran `ruff` and `black` manually on my code.

See [CONTRIBUTING.md](https://github.com/Azure-Samples/azure-search-openai-demo/blob/main/CONTRIBUTING.md#submit-pr) for more details.

en
es
fr
jp

### CODE REVIEW

**Comment by pamelafox** [comment]
Very cool, thanks for the PR!! I won't have any bandwidth to review PRs this week as we've got some live streams happening, but I'll take a look next week.

**Comment by bnodir** [comment]
@microsoft-github-policy-service agree company="NEC Corporation"

**Comment by bnodir** [comment]
Please review this PR as it is nearly complete. I will make any necessary revisions based on your feedback. Thank you.

**Comment by pamelafox** [comment]
Okay, I will review this week, thanks for letting me know it's ready!

File: app/frontend/package.json
Comment (pamelafox) [inline app/frontend/package.json:29]: The react-helmet is no longer maintained. The creator seems to recommend react-helmet-async instead, which gets a much better score on Snyk (https://snyk.io/advisor/npm-package/react-helmet-async)

File: app/frontend/src/components/Answer/SpeechOutputBrowser.tsx
Comment (pamelafox) [inline app/frontend/src/components/Answer/SpeechOutputBrowser.tsx:27]: Will this always find a matching voice? I'm worried it won't, and then there'll be an error. Could it default to en-US in that case?

File: app/frontend/src/i18n/formatters.tsx
Comment (pamelafox) [inline app/frontend/src/i18n/formatters.tsx:9]: Are these functions all expected by i18next? I can't find any direct use of them in the repository. Can you add a comment/link so that people don't accidentally delete them, thinking they're unused?

File: app/frontend/src/i18n/formatters.tsx
Comment (pamelafox) [inline app/frontend/src/i18n/formatters.tsx:7]: You have export in front of the functions and at the top, are both needed? (Maybe they are, I'm not a ES6 expert)

File: app/frontend/src/i18n/formatters.tsx
Comment (pamelafox) [reply app/frontend/src/i18n/formatters.tsx:9]: Hm, I'm reading https://www.i18next.com/translation-function/formatting but it kind of looks like these are the legacy formatters approach?

File: app/frontend/src/locales/es/translation.json
Comment (pamelafox) [inline app/frontend/src/locales/es/translation.json:2]: Thanks for the starter translations! Were these AI generated? Should I get a native speaker to review them?

File: app/frontend/src/pages/ask/Ask.tsx
Comment (pamelafox) [inline app/frontend/src/pages/ask/Ask.tsx:239]: Does this literally change the page title? If so, a comment might be helpful. It looks a bit funny in a div. Would it make sense in a higher-level component?

**Comment by pamelafox** [comment]
@bnodir This is so cool! I pushed one change to move the dropdown to a fluentUI dropdown and add a label, which makes the UI more consistent and fixes an accessibility warning.

We need to make the language picker optional, however, as not all developers will be able/willing to support translating all their strings, and some developers are targeting a single language. I think the way to do that is the same way we decide whether to show the vision options- modify the /config route to pass down a "showLanguagePicker" option, add an azd environment variable like ENABLE_LANGUAGE_PICKER, and set that accordingly.
Let me know if you need help implementing that.

Also, if you can, please compare the build times before this change and after this change. We always want to do that when adding new npm packages, so that we're aware of the performance hit of adding a particular package.

**Comment by pamelafox** [comment]
@zedhaque I think you were looking into i18n, potentially, for your audience? Am curious if you have any feedback on this PR. Also if you foresee issues with responsiveness.

**Comment by bnodir** [comment]
@pamelafox Thank you for reviewing the PR and providing your suggestions! I will work on your feedback this week and let you know once I am finished. Meanwhile, could I kindly ask you to have a native speaker review the translations, as they were generated by AI?

Regarding making the language picker optional and comparing build times, I will work on implementing the approach you suggested and may reach out if help is needed.

File: app/frontend/src/locales/fr/translation.json
Comment (pamelafox) [inline app/frontend/src/locales/fr/translation.json:61]:
```suggestion
        "promptTemplate": "Modèle de l'invitation",
```

File: app/frontend/src/locales/fr/translation.json
Comment (pamelafox) [inline app/frontend/src/locales/fr/translation.json:64]:
```suggestion
        "minimumRerankerScore": "Score minimum du reclasseur sémantique",
```

File: app/frontend/src/locales/fr/translation.json
Comment (pamelafox) [inline app/frontend/src/locales/fr/translation.json:67]:
```suggestion
        "useSemanticRanker": "Utiliser le reclasseur sémantique",
```

File: app/frontend/src/locales/fr/translation.json
Comment (pamelafox) [inline app/frontend/src/locales/fr/translation.json:97]:
```suggestion
        "shouldStream": "Diffuser en continu les réponses"
```

File: app/frontend/src/locales/fr/translation.json
Comment (pamelafox) [inline app/frontend/src/locales/fr/translation.json:104]:
```suggestion
            "Définit la température de la requête à le Grand modèle de langage (LLM) qui génère la réponse. Des températures plus élevées donnent lieu à des réponses plus créatives, mais elles peuvent être moins ancrées.",
```

**Review by vhvb1989 — COMMENTED** [review]
Reviewed Spanish translation and added small suggestions.

Great work, thank you!

**Comment by pamelafox** [comment]
@zedhaque I thought about that, but I think that the backend prompt language should probably be a separate knob, based on the language of the domain data versus the language of the user's questions. At least, the times that I recommend changing the prompt language are when developers are using completely non-English data. That said, those developers would likely also change the frontend language to default to non-English in that case. 

I could imagine DEFAULT_UI_LANGUAGE and DEFAULT_PROMPT_LANGUAGE in the future. I don't think we need that in this PR yet, unless it would be incompatible with the current design. Thoughts?

**Comment by bnodir** [comment]
@pamelafox I have completed the revisions based on your feedback. Could you please review it again?

Regarding comparing build times before and after the change, I am unsure how to do this. I have made changes to the following files, but I'm not entirely sure if this is correct. I will try to figure it out later. If you could provide any reference information, I would appreciate it.
- app/frontend/package-lock.json
- app/frontend/package.json
- package-lock.json

@vhvb1989 and @yoshioterada - Thank you very much for reviewing the translation files. I have made the corrections per your feedback. I would appreciate it if you could check it again.

File: app/frontend/package.json
Comment (bnodir) [reply app/frontend/package.json]: Thank you, changed to react-helmet-async.

File: app/frontend/src/components/Answer/SpeechOutputBrowser.tsx
Comment (bnodir) [reply app/frontend/src/components/Answer/SpeechOutputBrowser.tsx]: Thank you for the feedback. I have addressed the concern regarding defaulting to "en-US" if a matching voice for lngCode is not found.

File: app/frontend/src/i18n/formatters.tsx
Comment (bnodir) [reply app/frontend/src/i18n/formatters.tsx]: Thank you very much for your feedback. I have changed the code to use the built-in formatting functions instead of the legacy formatters, so this file is no longer needed.

File: app/frontend/src/i18n/formatters.tsx
Comment (bnodir) [reply app/frontend/src/i18n/formatters.tsx]: The file is no longer needed, and I have deleted it.

File: app/frontend/src/locales/fr/translation.json
Comment (bnodir) [reply app/frontend/src/locales/fr/translation.json]: Thank you, I have made the changes as per your suggestions.

File: app/frontend/src/locales/fr/translation.json
Comment (bnodir) [reply app/frontend/src/locales/fr/translation.json]: Thank you, I have made the changes as per your suggestions.

File: app/frontend/src/locales/fr/translation.json
Comment (bnodir) [reply app/frontend/src/locales/fr/translation.json]: Thank you, I have made the changes as per your suggestions.

File: app/frontend/src/locales/fr/translation.json
Comment (bnodir) [reply app/frontend/src/locales/fr/translation.json]: Thank you, I have made the changes as per your suggestions.

File: app/frontend/src/locales/fr/translation.json
Comment (bnodir) [reply app/frontend/src/locales/fr/translation.json]: Thank you, I have made the changes as per your suggestions.

File: app/frontend/src/locales/es/translation.json
Comment (bnodir) [reply app/frontend/src/locales/es/translation.json]: Thank you, I have made the changes as per your suggestions.

File: app/frontend/src/locales/es/translation.json
Comment (bnodir) [reply app/frontend/src/locales/es/translation.json]: Thank you, I have made the changes as per your suggestions.

**Comment by bnodir** [comment]
@zedhaque - Thank you for the clarification. Could you please check the attached screenshot to confirm whether the payload request looks correct? If you have any specific adjustments or suggestions, please let me know.
image

**Comment by john0isaac** [comment]
Yeah, all of the issues you mentioned are already fixed in the PR.
See my comment here:
https://github.com/Azure-Samples/azure-search-openai-demo/pull/1894#issuecomment-2277357613

I did say that they are out of the scope of the PR but went ahead and fixed all of them anyway.
Currently waiting for a review from a maintainer.

**Comment by pamelafox** [comment]
Okay, so is this PR ready for re-review, if all the commented issues are addressed?
And I'll also take a look at John's.

**Comment by bnodir** [comment]
@pamelafox - Yes, all the commented issues have been addressed, and it's ready for re-review. Kindly requesting your feedback.

To minimize the impact on build time, we can avoid adding the react-helmet-async npm package (which is used solely for changing the page title), as it seems to me to be of lesser importance.

**Comment by pamelafox** [comment]
I'd like to reduce the existing JS size before bringing this in, since this adds ~1MB of JS. I have a PR for that here: https://github.com/Azure-Samples/azure-search-openai-demo/pull/1947

Otherwise, this PR is looking very good. I'll resolve the merge conflicts.

**Review by pamelafox — APPROVED** [review]
Great work! Thanks for your patience on the review.

**Comment by bnodir** [comment]
Thank you so much!

**Pull request merged**
